File: src/edu/neu/musicId/Segmenter.java

package edu.neu.musicId;

import java.util.ArrayList;
import java.util.Arrays;

import edu.neu.musicId.util.Utilities;
import edu.neu.musicId.wav.WaveData;
import edu.neu.musicId.wav.WaveData.Chunk;
import edu.neu.musicId.wav.WaveDataFormat;

public class Segmenter {
    public double[][] segmentData(WaveData wave) {
        WaveDataFormat waveDataFormat1 = wave.getWaveDataFormat();

        Chunk chunk = wave.getChunk("data");

        double[] data = getSamples(wave, waveDataFormat1, chunk);

        // Computes the size of a float in bytes (answer: 4)
        final int FLOAT_SIZE = Float.SIZE / Byte.SIZE;

        // Compute the segmentation interval (in seconds)
        final double segmentInterval = getSegmentInterval(waveDataFormat1, chunk);

        // Compute the number of samples in each segment interval by:
        // segment length * byterate / (number of bytes per sample)
        int segmentInterval_numSamples = (int) (segmentInterval * waveDataFormat1.getByteRate())
                / waveDataFormat1.getBytesPerSample();

        // Compute the number of segments in this wave
        int numSegments = (int)(data.length / segmentInterval_numSamples);

        // Initialize array of arrays with dimensions:
        // Num Segments long x Num Samples per Segment wide
        double[][] segmentedData = new double[numSegments][segmentInterval_numSamples];

        // Declare temporary array and counters used by foreach loop
        double[] tempArray_doubles = new double[segmentInterval_numSamples];

        // TODO: Re-write for loop here using new getSamples() -Deniz

        System.out.println(segmentedData);

        return segmentedData;
    }

    private static double getSegmentInterval(WaveDataFormat format, Chunk chunk) {
        return 1F;
    }

    public static double[] getSamples(WaveData wave, WaveDataFormat format, Chunk chunk) {
        byte[][] channels = chunk.extractChannels(format);

        final int bytesPerSample = format.getBytesPerSample();
        final int blockAlign = format.getBlockAlign();
        final int numSamples = chunk.length() / blockAlign;
        final double divisor = Math.pow(2, 8 * bytesPerSample) - 1;

        double[] samples = new double[numSamples];

        for (int i = 0; i < samples.length; i++) {
            long sample = 0;

            for (int j = 0; j < channels.length; j++) {
                sample += Utilities.toInt(channels[j], (i * bytesPerSample), bytesPerSample,
                        (bytesPerSample < 2), wave.isLittleEndian());
            }
            
            samples[i] = (sample / (channels.length * divisor));
        }

        return samples;
    }
}

File: Segmenter.py
__author__ = 'Deniz'

from util import utilities

# duration of each segment of samples (in seconds)
SEGMENT_DURATION = 0.2

# sample rate to downsample audio to
TARGET_SAMPLE_RATE = 22050

# Segment the data in the given wav file
def segment_data(wave):
    format = wave.wave_data_format

    chunk = wave.chunk("data")

    downsample_factor = format.sample_rate / TARGET_SAMPLE_RATE

    sample_rate = format.sample_rate / downsample_factor
    byte_rate = format.byte_rate / downsample_factor

    samples_per_segment = int(SEGMENT_DURATION * byte_rate /
                              format.bytes_per_sample)

    num_samples = chunk.length() / format.block_align
    num_segments = num_samples / samples_per_segment / downsample_factor

    print "length: %d, block_align: %d" % (chunk.length(), format.block_align)
    print "sample_rate: %d, byte_rate: %d, downsample_factor: %d" % (sample_rate, byte_rate, downsample_factor)
    print "samples_per_segment: %d, num_samples: %d, num_segments: %d" % (samples_per_segment, num_samples, num_segments)

    if format.bytes_per_sample == 1:
        convert = utilities.to_byte
    elif format.bytes_per_sample == 2:
        convert = utilities.to_short
    else:
        convert = utilities.to_long

    is_unsigned = format.bytes_per_sample < 2

    time_step = float(samples_per_segment) / sample_rate

    segments = []

    for i in xrange(num_segments):
        samples = []

        for j in xrange(samples_per_segment / downsample_factor):
            sample = 0

            for k in xrange(format.num_channels):
                sample_offset = downsample_factor *\
                        (i * samples_per_segment + j)

                byte_offset = sample_offset * format.block_align +\
                        k * format.bytes_per_sample

                try:
                    sample += convert(chunk.data,
                                      byte_offset,
                                      is_unsigned,
                                      wave.is_little_endian)
                except:
                    print "sample_offset: %d, byte_offset: %d" % (sample_offset, byte_offset)
                    return

            samples.append(float(sample) / format.num_channels)

        segments.append((samples, samples_per_segment * time_step))

    return segments
